File: Semester2/2/task4/task4/LinkedStack.cs
using System;

namespace Task4
{
    /// <summary>
    /// Stack based on links
    /// </summary>
    class LinkedStack<T> : InterfaceStack<T>
    {
        private class StackElement
        {
            public StackElement(T value, StackElement next)
            {
                this.Value = value;
                this.Next = next;
            }

            public T Value { get; set; }
            public StackElement Next { get; set; }
        }

        private StackElement head;

        public void Push(T value)
        {
            StackElement newElement = new StackElement(value, head);
            head = newElement;
        }

        public T Pop()
        {
            if (IsEmpty())
                return default(T);
            T value = head.Value;
            head = head.Next;
            return value;
        }

        public T Top()
        {
            if (IsEmpty())
                return default(T);
            return head.Value;
        }

        public bool IsEmpty()
        {
            return (head == null);
        }

        public void PrintStack()
        {
            StackElement auxilary = head;
            while (auxilary != null)
            {
                Console.Write(auxilary.Value + " ");
                auxilary = auxilary.Next;
            }
            Console.WriteLine();
        }
    }
}

File: qt/Calc/linkedStack.cpp
#include "linkedStack.h"
#include "myException.h"

LinkedStack::LinkedStack() :
    mHead(nullptr)
  , mCurrentSize(0)
{
}

struct LinkedStack::StackElement
{
    double value;
    StackElement *next;
};

void LinkedStack::push(double value)
{
    StackElement *auxilary = new StackElement;
    auxilary->next = mHead;
    auxilary->value = value;
    mHead = auxilary;
    mCurrentSize++;
}

double LinkedStack::pop()
{
    if(size() == 0)
    {
        throw MyException("Stack is empty");
    }

    mCurrentSize--;
    double result = mHead->value;
    StackElement *auxilary = mHead;
    mHead = mHead->next;
    delete auxilary;
    return result;
}

double LinkedStack::top() const
{
    return mHead->value;
}

int LinkedStack::size() const
{
    return mCurrentSize;
}

LinkedStack::~LinkedStack()
{
    while(mHead != nullptr)
    {
        StackElement *auxilary = mHead;
        mHead = mHead->next;
        delete auxilary;
    }
    delete mHead;
}
